src/hash-table.h

```diff
@@ -235,15 +246,15 @@ void hash_table_iterate(HashTable *hash_table, HashTableIterator *iter);
 int hash_table_iter_has_more(HashTableIterator *iterator);
 
 /**
- * Using a hash table iterator, retrieve the next key.
+ * Using a hash table iterator, retrieve the next @ref KeyValuePair.
  *
  * @param iterator            The hash table iterator.
- * @return                    The next key from the hash table, or
+ * @return                    The next @ref KeyValuePair from the hash table, or
  *                            @ref HASH_TABLE_NULL if there are no more
  *                            keys to iterate over.
  */
 
-HashTableValue hash_table_iter_next(HashTableIterator *iterator);
+KeyValuePair *hash_table_iter_next(HashTableIterator *iterator);
```

Comment:
You are right. The table  internal structure will surely be a mess - if the pointed key is tempered with. 

I have therefore adopted the first approach i.e return a copy of  the `HashTablePair`.